move_base/src/move_base.cpp

```diff
@@ -560,6 +566,86 @@ namespace move_base {
     return global_pose;
   }
 
+  //publish feedback based on latest_plan.Called in context of executeCycle.
+  void MoveBase::publishFeedback(const geometry_msgs::PoseStamped& current_position)
+  {
+    //calculate distance to goal from current position
+    std::pair<bool, double> distance_info = calculateDistanceToGoal(current_position);
+    if(!distance_info.first)
+    {
+      return;
+    }
+    //calculate average velocity
+    ros::Time current_time = ros::Time::now();
+    double dist = distance_info.second;
+    calculateAverageVelocity(dist, current_time);
+
+    //publish feedback after X(30) iterations
+    if (prev_feedback_info_.average_velocity_iteration > 30) {
+      move_base_msgs::Feedback msg;
+      msg.dist_to_goal = dist;
+      msg.time_to_goal = prev_feedback_info_.calculateTimeToGoal(dist);
+      feedback_distance_pub_.publish(msg);
+    }
+
+    //save previous feedback info
+    prev_feedback_info_.prev_time = current_time;
+    prev_feedback_info_.dist_to_goal = dist;
+  }
+
+  //calculate distance from current position to the goal
+  std::pair<bool, double> MoveBase::calculateDistanceToGoal(const geometry_msgs::PoseStamped& current_position)
+  {
+    boost::unique_lock<boost::recursive_mutex> lock(planner_mutex_);
+    if(latest_plan_->empty())
+    {
+      return std::make_pair(false, 0.0);
+    }
+    std::vector<geometry_msgs::PoseStamped> plan;
+    plan = *latest_plan_;
+    lock.unlock();
+    //find the closest pose to current position, based on straigth line distance.
+    uint32_t min_index = 0;
+    double min_distance = distance(plan[0], current_position);
+    
+    for(uint32_t i = 1; i < plan.size(); ++i)
+    {
+      double current_distance = distance(plan[i], current_position);
+      if(current_distance < min_distance)
+      {
+        min_distance = current_distance;
+        min_index = i;
+      }
+    }
+
+    //calculate distance from closest pose to current position to the goal.
+    double dist = 0.0;
+    for(uint32_t idx = min_index; idx < plan.size() - 1; ++idx)
+    {
+      dist += distance(plan[idx], plan[idx + 1]);
+    }
+
+    return std::make_pair(true, dist);
+  }
+
+  //The average velocity calculation, based on "exponentially weighted moving average"(EWMA)
+  //with bias_correction.
+  void MoveBase::calculateAverageVelocity(double dist_to_goal, const ros::Time& current_time)
```

Comment:
I strongly recommend to make this a `FeedbackInfo` class's method. Doesn't make sense to keep it as a separate function - it's used by `FeedbackInfo` only. Now EWMA stuff is kept in `FeedbackInfo` class, and computed in this separate function - logically split. While in the global scope, `calculateAverageVelocity` can mislead other developers, bcz nothing says in its name it's for average velocity along the global path.

Reply:
thanks, I will fix it